### PR TITLE
(feature) Google SSO

### PR DESCRIPTION
**Pull request by Blaumaus**

### CODE REVIEW

File: src/auth/auth.controller.ts [hunk 1]
```diff
@@ -103,11 +105,12 @@ export class AuthController {
...
 
-    const jwtTokens = await this.authService.generateJwtTokens(newUser.id, true)
+    const { accessToken, refreshToken } =
+      await this.authService.generateJwtTokens(newUser.id, true)
 
     return {
-      accessToken: jwtTokens.accessToken,
-      refreshToken: jwtTokens.refreshToken,
+      accessToken,
+      refreshToken,
```
Comment (yevheniikulisidi) [inline src/auth/auth.controller.ts:113]: You can make `...jwtTokens`

File: src/auth/auth.controller.ts [hunk 2]
```diff
@@ -103,11 +105,12 @@ export class AuthController {
       body.password,
     )
 
-    const jwtTokens = await this.authService.generateJwtTokens(newUser.id, true)
+    const { accessToken, refreshToken } =
```
Comment (yevheniikulisidi) [inline src/auth/auth.controller.ts:108]: Make it like it was before (`jwtTokens`)

File: src/auth/auth.controller.ts [hunk 3]
```diff
@@ -412,4 +415,76 @@ export class AuthController {
...
+    return this.authService.processGoogleToken(token, hash)
+  }
+
+  @ApiOperation({ summary: 'Auth user' })
+  @Post('google/hash')
+  @Public()
+  // Validates the authorisation code and returns the JWT tokens
+  async getJWTByHash(
+    @Body() body: AuthUserGoogleDto,
+    @Headers() headers: unknown,
```
Comment (yevheniikulisidi) [inline src/auth/auth.controller.ts:449]: You can typify that.

File: src/auth/auth.service.ts [hunk 4]
```diff
@@ -410,4 +445,201 @@ export class AuthService {
...
+
+    if (user.isTwoFactorAuthenticationEnabled) {
+      user = _pick(user, ['isTwoFactorAuthenticationEnabled', 'email'])
+    } else {
+      user = await this.getSharedProjectsForUser(user)
+    }
+
+    return {
+      accessToken: jwtTokens.accessToken,
+      refreshToken: jwtTokens.refreshToken,
```
Comment (yevheniikulisidi) [inline src/auth/auth.service.ts:474]: Make it like it was before (`jwtTokens`)

File: src/auth/dtos/google-auth-process-code.dto.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { ApiProperty } from '@nestjs/swagger'
+import { IsString, IsUUID } from 'class-validator'
+
+export class AuthUserGoogleProcessCodeDto {
+  @ApiProperty({
+    description: 'Google auth token',
+    example: '4/0AVHE...',
+  })
+  @IsString()
+  token: string
```
Comment (yevheniikulisidi) [inline src/auth/dtos/google-auth-process-code.dto.ts:10]: Set the maximum length.

File: src/user/dto/user-google.dto.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  registeredWithGoogle: boolean
+
+  @ApiProperty({ example: '2023-01-01 23:12:55', required: true })
+  @IsString()
+  trialEndDate: string
+
+  @ApiProperty({ example: 'hello@example.com', required: false })
+  @IsEmail()
+  email?: string
+}
```
Comment (yevheniikulisidi) [inline src/user/dto/user-google.dto.ts:20]: Add checks https://github.com/typestack/class-validator#validation-decorators

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/user/dto/user-google.dto.ts (hunk 6 above)
Comment (Blaumaus) [reply src/user/dto/user-google.dto.ts:20]: it's internal dto so no need for advanced checks

**Pull request merged**